### DISCUSSION TITLE
What is the point of bundles

### DISCUSSION THREAD

**Discussion by haasal** [bevyengine/bevy#5051] [Q&A] [answered]
I don't quite understand what bundles are for.
Why should I use a bundle and query all components like transforms etc. separately instead of simply including them all inside a struct like this:

```rust
#[derive(Component)]
struct Object {
    transform: Transform,
    ...
}
```

**Comment by alice-i-cecile** [comment, marked as answer]
Great question! There are two good reasons to use bundles + smaller components over a single large component:

1. Systems automatically detect and report the data they require access to at the component (well, archetype / resource) level. Compatible systems can automatically run in parallel. By sticking your data into large components, systems that want to operate on your object must wait on each other, even if they're accessing different fields.
2. By splitting your data into small components, you can make systems that apply across object types. Rather than having a gravity system for every type of object, you can search for the `AffectedByGravity` marker component + the `Transform` component and generalize your logic in a very natural way.

**Reply by haasal** [reply]
Thanks. I think I understand :)

**Reply by haasal** [reply]
Ok but I just noticed that bundling makes it really hard defining methods. For example if I wanted to implement a `move(x: f32)` function on my Object I can't do that because only the ObjectBundle has access to the transform. However I cannot query the bundle directly in a system.

```rust
#[derive(Bundle)]
struct ObjectBundle {
    transform: Transform
    object: Object,
}
```

**Reply by alice-i-cecile** [reply]
There are two options here:

1. Custom system parameters, using the `#[derive(SystemParam)]` macro.
2. Use free floating functions that take in e.g. `&mut Transform` and work directly on component data.
